lib/PrintAsObjC/PrintAsObjC.cpp

```diff
@@ -912,62 +911,109 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
     }
     return hasPrintedAnything;
   }
-    
-  void printRenameForDecl(const AvailableAttr *AvAttr, const ValueDecl *D,
-                          bool includeQuotes) {
-    assert(!AvAttr->Rename.empty());
-    
-    auto renamedParsedDeclName = parseDeclName(AvAttr->Rename);
-    auto renamedDeclName = renamedParsedDeclName.formDeclName(D->getASTContext());
-    
+
+  const ValueDecl *getRenameDecl(const ValueDecl *D,
+                                 const ParsedDeclName renamedParsedDeclName) {
     auto declContext = D->getDeclContext();
-    const ValueDecl *renamedDecl = nullptr;
-    
+    ASTContext &astContext = D->getASTContext();
+    auto renamedDeclName = renamedParsedDeclName.formDeclName(astContext);
+
     if (isa<ClassDecl>(D) || isa<ProtocolDecl>(D)) {
       UnqualifiedLookup lookup(renamedDeclName.getBaseIdentifier(),
-                               declContext->getModuleScopeContext(),
-                               nullptr,
+                               declContext->getModuleScopeContext(), nullptr,
                                SourceLoc(),
                                UnqualifiedLookup::Flags::TypeLookup);
-      renamedDecl = lookup.getSingleTypeResult();
+      return lookup.getSingleTypeResult();
     } else {
+      TypeDecl *typeDecl = declContext->getSelfNominalTypeDecl();
+
+      if (!renamedParsedDeclName.ContextName.empty()) {
+        auto contextIdentifier =
+            astContext.getIdentifier(renamedParsedDeclName.ContextName);
+        UnqualifiedLookup specificTypeLookup(
+            contextIdentifier, declContext->getModuleScopeContext(), nullptr,
+            SourceLoc(), UnqualifiedLookup::Flags::TypeLookup);
+        if (!specificTypeLookup.getSingleTypeResult()) {
+          return nullptr;
+        }
+      }
```

Comment:
I think what I want to do here is to check if the ContextName is valid and can be found, if not then we'll return null immediately

Reply:
Just because we can find it doesn't mean it's the same context, though, right? What if I rename `Int.magnitude` to `Double.binade`?

…I mean, hopefully no one does this anyway, and I don't think we're handling it properly in Swift's own fix-it, but it'd be worse to suggest the *wrong* thing, right?

Reply:
What I did here is to get a context name and search for a type of that context. So if it is renamed to `Int.magnitude` then this would find a type result of `Int` and then it will continue to do a printing

This is for validating the the context is OK to use not to compare to the current one.

So should I remove this?

Reply:
I'm not sure. Do you think we should trust that a context is sensible if the user writes it, or not trust any contexts at all?

Reply:
That’s a hard question. My initial thought is that it would be great if we check the context here since Xcode doesn’t autocomplete the name in the rename attributes and I typed a name wrong for a few times.

Reply:
I see. So we should remove this check here?

Reply:
It’s pretty late in my tome zone but I want to make sure that I understand it correctly. 

So you’re suggesting that we don’t need to lookup about the information of Context for now, do I understand your point correctly?

Reply:
I think I'm leaning towards:

```
if renamed decl is a type {
  if the rename string has a "context" part {
    give up and use the rename string as is
  } else {
    do UnqualifiedLookup (as you have here)
  }
} else {
  do qualified lookup (as you have here), whether or not there's context
}
```

* * *

I also noticed that the `isa<ClassDecl> || isa<ProtocolDecl>` should also include `isa<EnumDecl>` (which also deserves a test). At this point that could be a separate PR if you want, though.

Reply:
Got it. Thanks. I'll update the PR with your suggestion.

On the including of checking for enum type, I think I will do it in this PR though if you don't mind

Reply:
@jrose-apple I was working on the Enum case as you suggested and then I found that Swift doesn't print the availability declaration for enum yet.

Should we add it here or in another PR?

Reply:
Oof. Let's make that another PR, then.